File: src/uFire_PAR.cpp
#include "uFire_PAR.h"

uFire_PAR::uFire_PAR(uint8_t i2c_address)
{
  _address = i2c_address;
  #ifndef ARDUINO_SAMD_ZERO
  Wire.begin();
  #endif // ifndef ARDUINO_SAMD_ZERO
}

uFire_PAR::uFire_PAR()
{
  _address = 0x3B;
  #ifndef ARDUINO_SAMD_ZERO
  Wire.begin();
  #endif // ifndef ARDUINO_SAMD_ZERO
}

#ifdef ESP32
uFire_PAR::uFire_PAR(uint8_t sda, uint8_t scl, uint8_t i2c_address)
{
  _address = i2c_address;
  Wire.begin(sda, scl, 100000);
}

uFire_PAR::uFire_PAR(uint8_t sda, uint8_t scl)
{
  _address = 0X3B;
  Wire.begin(sda, scl, 100000);
}

#endif // ifndef ESP32

uFire_PAR::~uFire_PAR()
{}

float uFire_PAR::measurePAR()
{
  _send_command(PAR_MEASURE_PAR);
  delay(PAR_MEASUREMENT_TIME);
  ppfd = _read_register(PAR_PAR_REGISTER);

  return ppfd;
}

uint8_t uFire_PAR::getVersion()
{
  return _read_byte(PAR_VERSION_REGISTER);
}

uint8_t uFire_PAR::getFirmware()
{
  return _read_byte(PAR_FW_VERSION_REGISTER);
}

void uFire_PAR::setI2CAddress(uint8_t i2cAddress)
{
  _write_register(PAR_BUFFER1_REGISTER, i2cAddress);
  _send_command(PAR_I2C);
  _address = i2cAddress;
}

bool uFire_PAR::connected()
{
  uint8_t retval;

  retval = _read_byte(PAR_VERSION_REGISTER);
  if (retval != 0xFF) {
    return true;
  }
  else {
    return false;
  }
}

float uFire_PAR::readEEPROM(uint8_t address)
{
  _write_register(PAR_BUFFER1_REGISTER, address);
  _send_command(PAR_READ);
  return _read_register(PAR_BUFFER2_REGISTER);
}

void uFire_PAR::writeEEPROM(uint8_t address, float value)
{
  _write_register(PAR_BUFFER1_REGISTER, address);
  _write_register(PAR_BUFFER2_REGISTER,   value);
  _send_command(PAR_WRITE);
}

void uFire_PAR::_change_register(uint8_t r)
{
  Wire.beginTransmission(_address);
  Wire.write(r);
  Wire.endTransmission();
  delay(10);
}

void uFire_PAR::_send_command(uint8_t command)
{
  Wire.beginTransmission(_address);
  Wire.write(PAR_TASK_REGISTER);
  Wire.write(command);
  Wire.endTransmission();
  delay(10);
}

void uFire_PAR::_write_register(uint8_t reg, float f)
{
  uint8_t b[5];
  float   f_val = f;

  b[0] = reg;
  b[1] = *((uint8_t *)&f_val);
  b[2] = *((uint8_t *)&f_val + 1);
  b[3] = *((uint8_t *)&f_val + 2);
  b[4] = *((uint8_t *)&f_val + 3);
  Wire.beginTransmission(_address);
  Wire.write(b, 5);
  Wire.endTransmission();
  delay(10);
}

float uFire_PAR::_read_register(uint8_t reg)
{
  float retval;

  _change_register(reg);
  Wire.requestFrom(_address, (uint8_t)1);
  *((uint8_t *)&retval) = Wire.read();
  Wire.requestFrom(_address, (uint8_t)1);
  *((uint8_t *)&retval + 1) = Wire.read();
  Wire.requestFrom(_address, (uint8_t)1);
  *((uint8_t *)&retval + 2) = Wire.read();
  Wire.requestFrom(_address, (uint8_t)1);
  *((uint8_t *)&retval + 3) = Wire.read();
  delay(10);
  return retval;
}

void uFire_PAR::_write_byte(uint8_t reg, uint8_t val)
{
  uint8_t b[5];

  b[0] = reg;
  b[1] = val;
  Wire.beginTransmission(_address);
  Wire.write(b, 2);
  Wire.endTransmission();
  delay(10);
}

uint8_t uFire_PAR::_read_byte(uint8_t reg)
{
  uint8_t retval;

  _change_register(reg);
  Wire.requestFrom(_address, (uint8_t)1);
  retval = Wire.read();
  delay(10);
  return retval;
}

File: python/RaspberryPi/uFire_PAR.py
import math
import struct
import time
import smbus  # pylint: disable=E0401

global i2c
UFIRE_PAR = 0x3B

PAR_MEASURE_PAR = 80
PAR_I2C = 40
PAR_READ = 20
PAR_WRITE = 10

PAR_VERSION_REGISTER = 0
PAR_FW_VERSION_REGISTER = 1
PAR_PAR_REGISTER = 2
PAR_BUFFER1_REGISTER = 6
PAR_BUFFER2_REGISTER = 10
PAR_TASK_REGISTER = 14

PAR_I2C_ADDRESS_REGISTER = 200

PAR_MEASUREMENT_TIME = 250


class uFire_PAR(object):
    ppfd = 0

    def __init__(self, address=UFIRE_PAR, i2c_bus=3, **kwargs):
        global i2c
        self.address = address
        i2c = smbus.SMBus(i2c_bus)

    def measurePAR(self):
        self._send_command(PAR_MEASURE_PAR)
        time.sleep(PAR_MEASUREMENT_TIME / 1000.0)
        self.ppfd = self._read_register(PAR_PAR_REGISTER)
        return self.ppfd

    def getVersion(self):
        return self._read_byte(PAR_VERSION_REGISTER)

    def getFirmware(self):
        return self._read_byte(PAR_FW_VERSION_REGISTER)

    def setI2CAddress(self, i2cAddress):
        if i2cAddress >= 1 and i2cAddress <= 127:
            self._write_register(PAR_BUFFER2_REGISTER, float(i2cAddress))
            self._send_command(PAR_I2C)
            self.address = int(i2cAddress)

    def connected(self):
        retval = self._read_byte(PAR_VERSION_REGISTER)

        if retval != 0xFF:
            return True
        else:
            return False

    def readEEPROM(self, address):
        self._write_register(PAR_BUFFER1_REGISTER, int(address))
        self._send_command(PAR_READ)
        return self._read_register(PAR_BUFFER2_REGISTER)

    def writeEEPROM(self, address, val):
        self._write_register(PAR_BUFFER1_REGISTER, int(address))
        self._write_register(PAR_BUFFER2_REGISTER, float(val))
        self._send_command(PAR_WRITE)

    def _bit_set(self, v, index, x):
        mask = 1 << index
        v &= ~mask
        if x:
            v |= mask
        return v

    def _change_register(self, r):
        global i2c
        i2c.write_byte(self.address, r)
        time.sleep(10 / 1000.0)

    def _send_command(self, command):
        global i2c
        i2c.write_byte_data(self.address, PAR_TASK_REGISTER, command)
        time.sleep(10 / 1000.0)

    def _write_register(self, reg, f):
        global i2c
        n = self.round_total_digits(f)
        fd = bytearray(struct.pack("f", n))
        data = [0, 0, 0, 0]
        data[0] = fd[0]
        data[1] = fd[1]
        data[2] = fd[2]
        data[3] = fd[3]
        self._change_register(reg)
        i2c.write_i2c_block_data(self.address, reg, data)
        time.sleep(10 / 1000.0)

    def _read_register(self, reg):
        global i2c
        data = [0, 0, 0, 0]
        self._change_register(reg)
        data[0] = i2c.read_byte(self.address)
        data[1] = i2c.read_byte(self.address)
        data[2] = i2c.read_byte(self.address)
        data[3] = i2c.read_byte(self.address)
        ba = bytearray(data)
        f = struct.unpack('f', ba)[0]
        return self.round_total_digits(f)

    def _write_byte(self, reg, val):
        global i2c
        i2c.write_byte_data(self.address, reg, val)
        time.sleep(10 / 1000.0)

    def _read_byte(self, reg):
        global i2c
        self._change_register(reg)
        time.sleep(10 / 1000.0)
        return i2c.read_byte(self.address)

    def magnitude(self, x):
        if math.isnan(x):
            return 0
        return 0 if x == 0 else int(math.floor(math.log10(abs(x)))) + 1

    def round_total_digits(self, x, digits=7):
        return round(x, digits - self.magnitude(x))
